Initial import

dbungert opened the pull request:
Enough structure to demonstrate chain-boot to ISOs of a different series.

Known issues:
* [x] needs all the error checking
* [x] casper-getty is a thing, do I need agetty?
  * casper-getty seems to be a removed feature from casper, and won't help.
* [ ] investigation needed on pmem region alignment, it seems to be either 2 or 4 MiB and 4MiB is done currently
  * later
* [ ] casper arguments such as a direct iso-url not implemented, and other arguments are either initially interesting or should be sent to kexec for later boots
  * later
* [ ] dmesg -c is not enough and we still get junk on the screen (also a Subiquity problem)
  * has a workaround which isn't great and won't work on serial 
* [ ] ncurses menu
  * [ ] move logging to syslog
    * done but maybe not helpful?  at least it isn't creating a standalone log now, easy to tweak later
  * [ ] json handling is very hardcoded, needs much more error checking and generalization
    * later
  * [ ] assumes ubuntu server
    * later
  * [ ] assumes things about color support of the terminal and needs a plan for lesser-capable color support
    * later

dbungert committed:
Initial import

ogayot commented on Makefile:
```diff
@@ -0,0 +1,37 @@
+CFLAGS+=$(shell pkg-config --cflags json-c)
+LDFLAGS+=$(shell pkg-config --libs json-c)
+
+SRCS:=$(shell echo *.c)
```
Minor, you can use the builtin `wildcard` function:
```make
SRCS:=$(wildcard *.c)
```

dbungert replied:
Fixed, thanks!
(And please keep the feedback coming, even if you think it's minor)

ogayot commented on main.c:
```diff
@@ -0,0 +1,440 @@
+
+void usage(char *prog)
+{
+    printf("usage: %s --input path --output path", prog);
```
Should we add a trailing `'\n'` ?

Also would `fprintf(stderr, ...)` make sense here?

dbungert replied:
Done, thanks.

ogayot commented on main.c:
```diff
@@ -0,0 +1,440 @@
+int text_white = COLOR_WHITE;
+int back_green = COLOR_GREEN;
+
+void usage(char *prog)
```
`noreturn`? I'm assuming we're buildling with >= C11.

dbungert replied:
Done, thanks.

ogayot commented on main.c:
```diff
@@ -0,0 +1,440 @@
+    free(args);
+}
+
+void logger(char *fmt, ...)
```
If you don't mind adding `__attribute__`, there is `format(printf ...)` which can help here.

dbungert replied:
Fixed by way of removing this in favor of syslog usage.

ogayot commented on main.c:
```diff
@@ -0,0 +1,440 @@
+
+    FILE *out = fopen("/tmp/iso-chooser-menu.log", "a");
+    vfprintf(out, fmt, ap);
+    fprintf(out, "\n");
```
minor, `fputc`?

dbungert replied:
Fixed by way of removing this in favor of syslog usage.

ogayot commented on main.c:
```diff
@@ -0,0 +1,440 @@
+    iso_data_t *ret = calloc(sizeof(iso_data_t), 1);
+    if(!ret) return NULL;
+
+    ret->label = label;
```
It feels weird to me to steal `label` and `url` from the caller while also expecting the memory to be allocated using malloc/calloc. Would you consider `strdup`ing here instead of in the caller?

dbungert replied:
The one hardcoded case now is to be removed, so long term I expect that the caller has already allocated memory.  I think I prefer to document my memory assumptions here.

dbungert replied:
Could also do something like this:
```
    choices->values[1] = iso_data_create(
            "Server",
            json_object_get_string(title),
            json_object_get_string(codename)),
            json_object_get_string(path)),
            json_object_get_int(size));
```
and let iso_data_create() generate the strings, but in general I wanted the choices_t to own the memory so that choices_free() would handle it without issue.

mwhudson replied:
It feels super dirty but I think it would be a defensible choice for this program to never deallocate anything!

mwhudson replied:
(exit_group(2) hides all memory management sins!)

ogayot replied:
I'm happy to mark this discussion resolved if needed.

ogayot commented on main.c:
```diff
@@ -0,0 +1,440 @@
+
+    erase();
+    refresh();
+    endwin();
```
I wonder if we should pass these calls to atexit to avoid exiting mid execution with a broken terminal

dbungert replied:
done

ogayot commented on scripts/30mini-iso-menu:
```diff
@@ -0,0 +1,114 @@
+
+    . /out.vars
+
+    dev=$(blkid | grep ISOIMAGE | cut -d: -f1)
```
consider using blkid --match-token ?

dbungert replied:
That looks safer, thanks!

ogayot reviewed (commented):
I don't know how much of this code will change going forward, so some of my comments may not be worth investing time into.

dbungert committed:
menu: use atexit() for cleanup

dbungert committed:
menu: minor fixes

dbungert committed:
casper: nicer blikd usage

dbungert committed:
menu: use syslog

mwhudson commented on debian/mini-iso-tools.postinst:
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+update-initramfs -u
```
So I know this is what casper does (because I went and looked &lt;wink&gt;) but there is also a dpkg trigger for rebuilding the initramfs. And if you say "apt install casper mini-iso-tools" it would be nice if we only rebuild the initramfs once. Could you take a few minutes to look into this? (and we should change casper to use this too of course)

dbungert replied:
At recentish debhelper-compat, it handles this.
Fixing casper means moving debhelper-compat which means fixing systemd handling.
The mini-iso-tools part is implemented.

mwhudson commented on Makefile:
```diff
@@ -0,0 +1,37 @@
+URL:=http://cdimage.ubuntu.com/streams/v1/com.ubuntu.cdimage.daily:ubuntu-server.json
+
+CFLAGS:=-Wall -Werror -Wfatal-errors
+LDFLAGS:=-static
```
We don't statically link most executables in the initrd aiui but rather rely on copy_exec to copy the dependencies in. It doesn't look as if libncurses or libjson-c are in the initrd already but it might be interesting to compare the difference between including them vs including the static binary.

dbungert replied:
On early analysis, the copy_exec magic with a non-static binary does appear to be smaller (difference of 660kiB), will look further on how that works and likely switch to the dynamic binary.

dbungert replied:
Final number was 550kiB, still worth doing.
Implemented.

mwhudson commented on scripts/iso-menu-session:
```diff
@@ -0,0 +1,16 @@
+
+set -e
+
+dmesg -c
```
After poking around for some related stuff, we might want to run "dmesg -n 4" here too (which I think sets the log level to the equivalent of putting "quiet" on the kernel command line)

dbungert replied:
This turned out to be a hard problem.  I have implemented chvt games to work around this.  The short answer is that stdout prints before the agetty risk bleeding into the terminal.
I have hopes that the above works for the Subiquity version of the problem.

mwhudson commented on scripts/30mini-iso-menu:
```diff
@@ -0,0 +1,112 @@
+    # Download the real ISO to the reserved memory region, and kexec to that
+
+    echo "###########################################"
+    echo "iso_chooser_step2"
```
Probably want slightly more human friendly message here :-)

dbungert replied:
Done

mwhudson commented on scripts/30mini-iso-menu:
```diff
@@ -0,0 +1,112 @@
+
+    echo "Loading $MEDIA_LABEL ..."
+
+    pmem_directive="memmap=$(/usr/bin/get_memmap_directive $MEDIA_SIZE)"
```
This isn't run under set -e afaics, so a failure here won't halt the boot? It probably should?

dbungert replied:
Done

mwhudson commented on scripts/30mini-iso-menu:
```diff
@@ -0,0 +1,112 @@
+
+    /usr/sbin/agetty --skip-login \
+        --login-program /usr/bin/iso-menu-session \
+        tty1 linux
```
This whole area of things makes my head hurt tremendously but I'm not sure we want tty1 here? Because in at least some cases we'll want it to run on the serial console. do_interactive_netmount just reads and writes from stdin/stdout/stderr -- do we even need agetty?

dbungert replied:
I used agetty as part of getting a terminal that ncurses was happy with.  (I wonder if casper-getty would suffice?)
The rest, I don't know yet.

mwhudson replied:
I think you might be able to get away with passing - instead of tty1? Not sure, sorry.

dbungert replied:
That crashes agetty, actually.  The manpage documents some assumptions, perhaps those are not valid assumptions in this state.
Not implemented.

mwhudson commented on scripts/iso-menu-session:
```diff
@@ -0,0 +1,16 @@
+
+cdimage_json=/cdimage.json
+wget -O $cdimage_json \
+    http://cdimage.ubuntu.com/streams/v1/com.ubuntu.cdimage.daily:ubuntu-server.json
```
https please!!!

dbungert replied:
Done.  Another a similar scenario in the menu C code.

mwhudson commented on scripts/30mini-iso-menu:
```diff
@@ -0,0 +1,112 @@
+    fi
+
+    echo "download $URL"
+    wget "${URL}" -O "${target}"
```
I don't know how much it really makes things better but it might be nice if we could valid the sha256 from the simple streams data here. Maybe.

dbungert replied:
Done.

mwhudson commented on Makefile:
```diff
@@ -0,0 +1,37 @@
+
+URL:=http://cdimage.ubuntu.com/streams/v1/com.ubuntu.cdimage.daily:ubuntu-server.json
+
+CFLAGS:=-Wall -Werror -Wfatal-errors -std=c11
```
We shouldn't really override the default CFLAGS here I think. I don't speak Makefile well enough any more to know what the best practice here is though.

dbungert replied:
Done.
https://www.gnu.org/software/make/manual/html_node/Appending.html was helpful.
Using the system CFLAGS exposed usage of undefined behavior, which I fixed.

mwhudson commented on main.c:
```diff
@@ -0,0 +1,438 @@
+/*
```
Is it worth splitting the UI code and the "reading stuff out of streams data" code into separate files?

I do really wish there was a better alternative to doing all this in C.

mwhudson replied:
Actually is it worth splitting the UI code and "reading stuff out of streams data" into separate _programs_? Because then it might be possible to write sensible tests for the latter one (or it can just be unit tests of the C code I guess -- but tests of that would be nice)

dbungert replied:
As a separate program I would have to pass data around, I think it can be effectively unit tested just split out to a different file.  Willing to keep the distinct program option on the table if we need it.

mwhudson commented on main.c:
```diff
@@ -0,0 +1,438 @@
+        init_color_from_bytes(text_white, 0xFF, 0xFF, 0xFF);
+        init_color_from_bytes(back_green, 0x0E, 0x84, 0x20);
+    } else {
+        /* These are terminal 256 color codes, see
```
I guess we can change this to ioctl(PIO_CMAP) at some point

mwhudson commented on main.c:
```diff
@@ -0,0 +1,438 @@
+    json_object *items = json_object_object_get(date, "items");
+    json_object *iso = json_object_object_get(items, "iso");
+    json_object *path = json_object_object_get(iso, "path");
+    json_object *size = json_object_object_get(iso, "size");
```
Well yeah this isn't really acceptable as you know. At the very least it needs error checking!

dbungert replied:
I have an initial answer here that is better, and further refinements that I'll save for a follow-on.  This PR is big enough.

mwhudson reviewed (commented):
Well I think that's enough disconnected comments for now. It mostly looks fine and I'm impressed you managed to clone the subiquity look and feel to the extent you did! Let me know when you want me to take another look.

dbungert committed:
d/postinst: remove explicit update-initramfs

dh_installinitramfs handles this already with a better method, triggers.

dbungert committed:
debian: 2 lintian nits

dbungert committed:
casper: tweak messages

dbungert committed:
session: use https

dbungert committed:
menu: https

dbungert committed:
several: sh256sum verification

dbungert committed:
casper: handle fsck.mode=skip

dbungert committed:
Makefile: use system CFLAGS

dbungert committed:
menu: fix error handling of vasprintf

dbungert committed:
Makefile: stop using static build

copy_exec handles this, better

dbungert committed:
several: hooks restructure

The goal here is to use copy_exec to copy the menu in, but now the
recursive copy of other stuff looks weird, so just throw things into the
/usr/lib/mini-iso-tools dir and send them to their final locations
later.

dbungert committed:
menu: redraws more often

Redraws are cheap enough that I'd rather redraw on any input if there is
cruft on the screen.

dbungert committed:
memmap: update for error handling friendliness

Caller can now check error code for determining pass / fail of the
memmap choice.

dbungert committed:
casper: workaround messages glitch

Stdout prints before the agetty may bleed into the display.
Prints after agetty takes over are safe.
Arguably this means using agetty earlier in the session.
Do chvt tricks for now.

dbungert committed:
menu: split files

Split out some items.  The main goal is testability of the json parts.

dbungert committed:
test: setup c unit tests

dbungert committed:
test: json find_largest_subkey

dbungert committed:
test: reader

mwhudson commented on scripts/30mini-iso-menu:
```diff
@@ -26,16 +26,14 @@ iso_chooser_step1() {
-    echo "iso_chooser_step1"
-    echo "###########################################"
-
+    chvt 2  # the chvts work around messages bleeding into the agetty
```
Do we want this to work at all over serial? I don't really know the implications of that question but this isn't going to work i think...

dbungert replied:
I expect it won't, and want to investigate in a future PR.

dbungert committed:
test: add current ubuntu-server json

dbungert committed:
d/control: declare libcmocka-dev build-dep

mwhudson reviewed (approved):
I think this is good enough for an initial landing now, thanks! I guess you might want to get @ogayot to take another look too.

ogayot commented on Makefile:
```diff
@@ -0,0 +1,56 @@
+
+BIN:=iso-chooser-menu
+
+URL:=http://cdimage.ubuntu.com/streams/v1/com.ubuntu.cdimage.daily:ubuntu-server.json
```
another s/http/https/ maybe?

dbungert replied:
Done.  No reason not to.

ogayot commented on Makefile:
```diff
@@ -0,0 +1,56 @@
+
+URL:=http://cdimage.ubuntu.com/streams/v1/com.ubuntu.cdimage.daily:ubuntu-server.json
+
+CFLAGS+=-Wall -Werror -Wfatal-errors -std=c11 -I.
```
AFAIK `-I.` is more of a `CPPFLAGS` kind of thing usually, but that should work too.

I suspect that it's not actually needed though, or is it?

dbungert replied:
It is at the moment actually, due to how I have structured the build of the test of the menu.
I'd like to restructure that test build anyway and can drop the `-I.` at that time.

ogayot replied:
OK

ogayot commented on main.c:
```diff
@@ -0,0 +1,323 @@
+    }
+}
+
+void exit_cb()
```
void exit_cb(void) ?

dbungert replied:
Done.

ogayot commented on common.h:
```diff
@@ -0,0 +1,45 @@
+choices_t *choices_create(int len);
+void choices_free(choices_t *c);
+
+char *saprintf(char *fmt, ...);
```
`__attribute__((format(printf, 1, 2)))` ?

dbungert replied:
Done.  I haven't done a tone with attributes with C yet, it looks like this should be in the .h file right?

ogayot replied:
yep! thanks!

ogayot replied:
~~Should go before the function declaration though :)~~
```
__attribute__((format(printf, 1, 2)))
char *saprintf(char *fmt, ...);

ogayot replied:
nevermind, that works too :)

ogayot commented on main.c:
```diff
@@ -0,0 +1,323 @@
+{
+    FILE *f = fopen(fname, "w");
+    if(!f) {
+        syslog(LOG_ERR, "failed to open output file [%s]", fname);
```
I think including the errno description could help for diagnosis:
`"failed to open output file [%s]: %m"` ?

dbungert replied:
nice, didn't know about that one.

ogayot commented on scripts/checksum-device/checksum-device:
```diff
@@ -0,0 +1,15 @@
+infile_size=$2
+expected=$3
+
+actual=$(dd if=$infile iflag=count_bytes count=$infile_size \
```
I'd be extra careful with the quoting here because dd is scary :skull:  

`dd if="$infile" iflag=count_bytes count="$infile_size"`

dbungert replied:
dd is scary.  Implemented.

ogayot reviewed (commented):
A few more comments but overall LGTM

dbungert committed:
build: s/http/https

dbungert committed:
checksum: quoting for safety

dbungert committed:
menu: misc cleanup
